platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,148 @@
+3. Formats the result into useful console output
+4. Breaks the build on production, and logs the output on lower environments
+
+##### Accessibility Checking
+
+
+
+#### Pain points
+
+##### Content Writers
+
+Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day, and `vets-website` deploys to staging take 20+ minutes. To allow content writers to move faster, there is a partial deploy, which only includes static page changes (`vagov-content` and Drupal).
+
+Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive reactive way to learn if their content includes broken links or accessibility errors.
+
+##### Users
+
+Content-only deploys create an opportunity for end users to encounter broken links and/or accessibility errors from CMS content that wasn't validated before being deployed.
+
+##### FE Tools team
+
+Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with another team.
+
+### High Level Design
+
+Both the application and content builds need content validation.
+
+## Specifics
+
+### Detailed Design
+
+COMING SOON
+
+### Code Location
+
+The validation of the application content will remain in `vets-website`. The validation of the static content will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).
+
+### Testing Plan
+
+COMING SOON
+
+### Logging
+
+COMING SOON
+
+### Debugging
+
+COMING SOON
+
+### Caveats
+
+- Our [function to determine if a link is broken](https://github.com/department-of-veterans-affairs/vets-website/blob/e8b68850fc83fdae4386fe4ab392770dde38faee/src/site/stages/build/plugins/check-broken-links/helpers/isBrokenLink.js#L16) doesn't check if external links are broken.
+- Our [accessibility check](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/assets/js/execute-axe-check.js#L8) only renders a banner on the preview server.
+
+### Security Concerns
+
+There are no new security concerns with a separated build process.
+
+### Privacy Concerns
+
+There are no new privacy concerns with a separated build process.
+
+### Open Questions and Risks
+
+COMING SOON
+
+### Work Estimates
+
+COMING SOON
+
+### Alternatives
+
+#### Highest quality, but lowest speed
+
+If we could proactively validate CMS content before it's published, we could eliminate a source of downstream broken links and/or accessibility errors, which would improve the experience for end users and/or reduce the number of frontend deploys that require manual intervention. The downside of this approach would be figuring out how to add the content validation to the CMS, and/or adding friction to the content writers workflow.
```

Comment:
@cvalarida We added link checking to the CMS several months back On node save every link both internal and external are tested and a report generated for that node.   
![image](https://user-images.githubusercontent.com/5752113/83689439-0a64c680-a5bd-11ea-9e38-e5c855f6f78e.png)
A bright red box appears if there are any that are broken.
![image](https://user-images.githubusercontent.com/5752113/83689693-79dab600-a5bd-11ea-81bb-58755dd95bbe.png)

508 compliance is a little tougher because the front end templates do not match the cms templates,

Reply:
This checker even reaches out to the public website in case it finds pages that are not known to drupal.  

All contributed back to benefit Drupal's everywhere :)  
https://www.drupal.org/project/node_link_report

Reply:
It is on my roadmap to add link accessibilty checking to the link checker which would warn on an editor creating a link with no text in it.   This was a recent issue that blocked a content release a couple of weeks back.
https://www.drupal.org/project/node_link_report/issues/3145477

Reply:
@swirtSJW how come when we see reports of broken link they seem to originate from failing builds? 

If we're confident with Drupal's link validation, then I think we should just delete link validation from vets-website.

Reply:
@ncksllvn I think most now are either related to something being unpublished or moved to a new path.  The node link report warns if linking to something that is unpublished, but it does not / can not prevent it.

Reply:
Also, our link validation happens on a save in drupal.  It is not blocking, only reported.   It is up to editors to note and fix.